File: Advent of Code.FSharp/Challenge Solutions/Year 2025/ChallengeSolution02.fs
namespace Advent_of_Code.FSharp.Challenge_Solutions.Year_2025

open Advent_of_Code.FSharp.Utilities
open Advent_of_Code.Shared
open Advent_of_Code.Shared.Utilities

module ChallengeSolution02 =

    type Range =
        {
            Start: string
            End: string
        }
    
    type Solution(getInput: unit -> string list, writeLine: string -> unit) =
        
        let readRanges() : Range array =
            getInput()
            |> List.head
            |> fun input -> input.Split ','
            |> Array.map
                ^<| fun range -> range.Split '-'
                    |> Array.map (fun splitRange ->
                        {
                            Start = splitRange[0]
                            End = splitRange[1]
                        })
                    
        let getSingleSizeRanges (ranges: Range array) : Range array =
            
            let rec getSubRanges (start: string) (final: string) : Range list =
                match start.Length < final.Length with
                | false -> [{ Start = start; End = final }]
                | true ->
                    let newStart = "1" + String.replicate start.Length "0"
                    let subRangeEnd = String.replicate start.Length "9"
                    [{ Start = start; End = subRangeEnd }] @ getSubRanges newStart final        
            
            ranges
            |> Array.collect (fun range ->
                getSubRanges range.Start range.End
                |> Array.ofList)

        let asDigits (number: int64) : int64 list =
            let rec asDigits (number: int64) : int64 list =
                match number with
                | 0L -> []
                | _ ->
                    let digit = number % 10L
                    (asDigits (number / 10L)) @ [digit]
                    
            match number with
            | 0L -> [0L]
            | _ -> asDigits number 
                
        let rec toNumber (digits: int64 list) : int64 =
            match digits with
            | [] -> 0
            | d::ds ->
                let number = toNumber ds
                number * 10L + d
                
        let areHalvesEqual (number: int64) : bool =
            let digits = number |> asDigits
            
            match digits.Length % 2 with
            | 0 ->
                digits
                |> List.splitAt (List.length digits / 2)
                |> fun (h1, h2) -> (toNumber h1, toNumber h2)
                |> fun (h1, h2) -> h1 = h2
            | _ -> false
            
        let isMadeOfRepeatedSequences (number: int64) : bool =
            let chunkNumber (number: int64) (sequenceLength: int) : string array =
                number
                |> string
                |> Seq.chunkBySize sequenceLength
                |> Seq.map System.String
                |> Seq.toArray
                
            let digits = number |> asDigits
            
            [|1..(digits.Length / 2)|]
            |> Array.filter (fun seqLength -> digits.Length % seqLength = 0)
            |> Array.map (chunkNumber number)
            |> Array.exists (fun chunks -> chunks |> Array.forall (fun chunk -> chunk = chunks[0]))
        
        let computeInvalidIdSum (isInvalid: int64 -> bool) : int64 =
            readRanges()
            |> getSingleSizeRanges
            |> Array.map (fun range ->
                [|(range.Start |> int64)..(range.End |> int64)|]
                |> Array.filter isInvalid
                |> Array.sum)
            |> Array.sum
        
        member _.SolveFirstPart() =
            let invalidIdSum = computeInvalidIdSum areHalvesEqual
            writeLine $"Invalid ID sum: {invalidIdSum}"

        member _.SolveSecondPart() : unit =
            let invalidIdSum = computeInvalidIdSum isMadeOfRepeatedSequences
            writeLine $"Invalid ID sum: {invalidIdSum}"

    type ChallengeSolution02(console: IConsole, reader: ISolutionReader<ChallengeSolution02>) =
        inherit ChallengeSolution<ChallengeSolution02>(console, reader)

        let solution = Solution(reader.ReadLines >> List.ofArray, console.WriteLine)

        override this.SolveFirstPart() = solution.SolveFirstPart()
        override this.SolveSecondPart() = solution.SolveSecondPart()

File: Advent of Code/Challenge Solutions/Year 2025/ChallengeSolution02.cs
// Task: https://adventofcode.com/2025/day/2

using Advent_of_Code.Shared;
using Advent_of_Code.Shared.Utilities;

namespace Advent_of_Code.Challenge_Solutions.Year_2025;

public class ChallengeSolution02(IConsole console, ISolutionReader<ChallengeSolution02> reader)
    : ChallengeSolution<ChallengeSolution02>(console, reader)
{
    public override void SolveFirstPart()
    {
        var ranges = SplitRangesByDigitCount(ReadRanges());
        RemoveRangesWithOddCount(ranges);

        long invalidIdSum = 0;

        foreach (var range in ranges)
        {
            for (var i = long.Parse(range.Start); i <= long.Parse(range.End); i++)
            {
                if (IsFirstHalfEqualToTheSecondHalf(i))
                {
                    invalidIdSum += i;
                }
            }
        }

        _console.WriteLine($"Invalid ID sum: {invalidIdSum}");
    }

    public override void SolveSecondPart()
    {
        var ranges = SplitRangesByDigitCount(ReadRanges());

        long invalidIdSum = 0;

        foreach (var range in ranges)
        {
            for (var i = long.Parse(range.Start); i <= long.Parse(range.End); i++)
            {
                if (IsMadeOfRepeatedSequences(i))
                {
                    invalidIdSum += i;
                }
            }
        }

        _console.WriteLine($"Invalid ID sum: {invalidIdSum}");
    }

    private static bool IsFirstHalfEqualToTheSecondHalf(long number)
    {
        var digits = GetDigits(number);
        var digitCount = digits.Length;

        if (digitCount % 2 != 0)
        {
            return false;
        }

        var leftHalf = DigitArrayToNumber(digits[..(digitCount / 2)]);
        var rightHalf = DigitArrayToNumber(digits[(digitCount / 2)..]);

        return leftHalf == rightHalf;
    }

    private static bool IsMadeOfRepeatedSequences(long number)
    {
        var digits = GetDigits(number);
        var digitCount = digits.Length;

        for (var sequenceLength = 1; sequenceLength <= digitCount / 2; sequenceLength++)
        {
            if (digitCount % sequenceLength != 0)
            {
                continue;
            }

            var sequences = ChunkString(number.ToString(), sequenceLength).ToArray();
            if (sequences[1..].All(seq => seq == sequences[0]))
            {
                return true;
            }
        }

        return false;
    }

    private static List<Range> SplitRangesByDigitCount(List<Range> ranges)
    {
        List<Range> newRangeList = [];

        foreach (var range in ranges)
        {
            var start = range.Start;
            var end = range.End;

            while (start.Length < end.Length)
            {
                var smallRangeEnd = new string('9', start.Length);
                newRangeList.Add(new(start, smallRangeEnd));

                start = $"1{new string('0', start.Length)}";
            }

            newRangeList.Add(new(start, end));
        }

        return newRangeList;
    }

    private static void RemoveRangesWithOddCount(List<Range> ranges) => ranges.RemoveAll(range => range.Start.Length % 2 != 0);

    private static IEnumerable<string> ChunkString(string @string, int size)
    {
        for (var i = 0; i < @string.Length; i += size)
        {
            var x = "";
            try
            {
                x = @string.Substring(i, size);
            }
            catch (Exception ex)
            {
                throw ex;
            }

            yield return x;
        }
    }

    private static int[] GetDigits(long number)
    {
        List<int> digitList = [];
        do
        {
            digitList.Add((int)(number % 10));
            number /= 10;
        } while (number > 0);

        return digitList.ToArray();
    }

    private static long DigitArrayToNumber(int[] array)
    {
        long number = 0;
        foreach (var x in array)
        {
            number = (number * 10) + x;
        }

        return number;
    }

    private List<Range> ReadRanges()
    {
        var input = _reader.ReadLines()[0];
        var rangeStrings = input.Split(',', StringSplitOptions.TrimEntries);

        List<Range> ranges = [];

        foreach (var range in rangeStrings)
        {
            var rangeElements = range.Split('-', StringSplitOptions.TrimEntries);
            var start = rangeElements[0];
            var end = rangeElements[1];

            ranges.Add(new(start, end));
        }

        return ranges;
    }

    private record struct Range(string Start, string End);
}
